src/Time/TimeSteppers/RungeKutta4.hpp

```diff
@@ -0,0 +1,218 @@
+      // time = t^n + dt
+      *u = U0 + 0.5 * (time_step.value() * dt_vars);
+      // On exit v = v^(2), time = t^n + (1/2)*dt
+      break;
+    }
+    case 2: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(3) = u^n + dt*RHS(v^(2), t^n + (1/2)*dt))
+      // On entry V = v^(2), U0 = u^n, rhs0 = RHS(v^(2), t^n + (1/2)*dt),
+      // time = t^n + (1/2)*dt
+      *u = U0 + time_step.value() * dt_vars;
+      // On exit v = v^(3), time = t^n + dt
+      break;
+    }
+    case 3: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // u^(n+1) = (2v^(1) + 4*v^(2) + 2*v^(3) + v^(4) - 3*U0)/6
+      // On entry V = v^(3), U0 = u^n, rhs0 = RHS(v^(3), t^n + dt),
+      // time = t^n + dt
+      // Note: v^(4) = U0 + dt*RHS(t+dt, v^(3)); inserting this gives
+      // u^(n+1) = (2v^(1) + 4*v^(2) + 2*v^(3) + dt*RHS(t+dt,v^(3)) - 2*U0)/6
+      constexpr double one_sixth = 1.0 / 6.0;
+      *u = (2.0 * (history->begin() + 1).value() +
+            4.0 * (history->begin() + 2).value() +
+            2.0 * (history->begin() + 3).value() +
+            (time_step.value() * dt_vars - 2.0 * U0)) *
+           one_sixth;
+      // On exit v = u^(n+1), time = t^n + dt
+      break;
+    }
+    default:
+      ERROR("Bad substep value in RK4: " << substep);
+  }
+
+  // Clean up old history
+  if (history->size() == number_of_substeps()) {
+    history->mark_unneeded(history->end());
+  }
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::dense_update_u(const gsl::not_null<Vars*> u,
+                                 const History<Vars, DerivVars>& history,
+                                 const double time) const noexcept {
+  ASSERT(history.size() == number_of_substeps(),
+         "Can only dense output on last substep");
+  const double step_start = history[0].value();
+  const double step_end = history[history.size() - 1].value();
+  const double time_step = step_end - step_start;
+  const double output_fraction = (time - step_start) / time_step;
+  ASSERT(output_fraction >= 0, "Attempting dense output at time "
+                                   << time << ", but already progressed past "
+                                   << step_start);
+  ASSERT(output_fraction <= 1, "Requested time ("
+                                   << time << ") not within step ["
+                                   << step_start << ", " << step_end << "]");
+
+  // Numerical Recipes Eq. (17.2.15). This implements cubic interpolation
+  // throughout the step. Because the history only is available through the
+  // penultimate step, i) the value after the step is computed algebraically
+  // from previous substeps, and ii) the derivative at the final step is
+  // approximated as the derivative at the penultimate substep.
+  constexpr double one_sixth = 1.0 / 6.0;
+  const auto& U0 = history.begin().value();
+  const auto& dt_vars = (history.end() - 1).derivative();
+  const auto& U1 =
```

Comment:
Why is this a reference? I'm also worried that this will create an expression template when using `auto` and so you are recomputing this expression twice below (if I counted correctly)

Reply:
I can't think of a reason not to just change to const auto u1

Reply:
I think `const auto u1` will mean the type is an expression and so would be re-evaluated twice below rather than computed once and stored. Does that make sense?